webdriver-spec.html

```diff
@@ -4624,636 +4646,562 @@
+      <a>action chains</a>. This is the equivalent of releasing all input sources.
+    <p>The <a>remote end steps</a> are:
+      <ol>
+
+        <li><p class=issue>This needs to be completed.
+
+        <li><p>Return <a>success</a> with data null.
+      </ol>
+    </section>
+    To be able to <dfn>dispatch commands</dfn> on the remote end we need
+    to perform the following steps:
     <ol>
-      <li>Generate a <code>keydown</code> event of <var>modifier</var></li>
+      <li>Let <var>command</var> be the value from <a>getting a property</a>
+        <code>name</code>.
+      <li>Match <var>command</var> against <a>Keyboard</a> or <a>Pointer</a>
+        actions, passing in values on <code>parameters</code> JSON Object.
     </ol>
-  </li>
-  <li>Maintain this key state and use it to modify the input until it is pressed again.</li>
-</ol>
-      </li>
-      <li>Else if <var>key</var> represents the NULL key:
+  <section> <!-- Commands -->
+    <h2>Action Commands</h2>
+    <section><!-- General Actions -->
+      <h2><dfn>General Actions</dfn></h2>
+      <p>The <dfn>pause</dfn> command:
+        <ol>
+          <li><p>If the <a>current top-level browsing context</a> is <a>no longer open</a>
+            , return <a>error</a> with <a>error code</a> <a>no such window</a>.
+
+          <li><p><a>Handle any user prompts</a> and return its value if it is an
+            <a>error</a>.
+
+          <li><p>Let <var>duration</var> be the result of <a>getting a property</a>
+            named "<code>duration</code>".
+
+          <li><p>If <var>duration</var> is not a <a>Number</a> return <a>error</a>
+            with <a>error code</a> <a>invalid argument</a>.
+
+          <li><p>Wait until <var>duration</var> milliseconds have passed.
+
+        </ol>
+    </section><!-- /General Actions -->
+
+    <section> <!-- Keyboard Actions -->
+    <h2><dfn data-lt=keyboard>Keyboard Actions</dfn></h2>
+    <p>The <dfn>keyUp</dfn> command:
       <ol>
-        <li>Generate <code>keyup</code> events of all modifier keys currently held
-        down.</li>
-        <li>All modifier keys are now assumed to be released.</li>
+        <li><p>If the <a>current top-level browsing context</a> is <a>no longer open</a>
+          , return <a>error</a> with <a>error code</a> <a>no such window</a>.
+
+        <li><p><a>Handle any user prompts</a> and return its value if it is an
+          <a>error</a>.
+
+        <li><p>Let <var>key</var> be the result of <a>getting a property</a>
+          named "<code>value</code>".
+
+        <li><p>Let <var>active element</var> be the <a>active element</a> of
+          the <a>current browsing context</a>’s <a>document element</a>.
+
+        <li><p>Let <var>key to type</var> be the result of <a>processing keys</a>
+          with argument <var>key</var>.
+
+        <li><p>Generate <a data-lt=keyUp-Event>keyUp</a> <a>Keyboard Event</a> against
+          <var>active element</var> for <var>key to type</var>.
```

Comment:
WebDriver does this. The client APIs find the API and do some work before doing the action. See https://github.com/SeleniumHQ/selenium/blob/master/py/selenium/webdriver/common/action_chains.py#L74

However... having atomic actions might make this extra work impossible... will have another think about it...